### PR TITLE
Issue 184

### PR DESCRIPTION
**Pull request by thozza**
- Extended tests for DownloadHelper
- replaced pycurl by pure Python implementation - pycurl is not needed any more
- fixed issue #184

### CODE REVIEW

File: rebasehelper/utils.py [hunk 1]
```diff
@@ -227,34 +234,46 @@ def progress(download_total, downloaded, upload_total, uploaded):
...
-            curl.setopt(pycurl.FOLLOWLOCATION, 1)
-            curl.setopt(pycurl.MAXREDIRS, 5)
-            curl.setopt(pycurl.TIMEOUT, 300)
-            curl.setopt(pycurl.WRITEDATA, f)
-            curl.setopt(pycurl.NOPROGRESS, 0)
-            curl.setopt(pycurl.PROGRESSFUNCTION, progress)
-            try:
+        try:
+            response = urllib.request.urlopen(url, timeout=timeout)
+            file_size = int(response.info().get('Content-Length', 0))
```
Comment (phracek) [inline rebasehelper/utils.py]: Does it work also for PyPi and other stuff?

File: rebasehelper/utils.py [hunk 2]
```diff
@@ -227,34 +234,46 @@ def progress(download_total, downloaded, upload_total, uploaded):
...
+
+            if not file_size:
+                raise DownloadError('The file has zero size')
+
+            # file exists, check the size
+            if os.path.exists(destination_path):
+                if file_size != os.path.getsize(destination_path):
+                    logger.debug("The destination file '%s' exists, but sizes don't match! Removing it.",
```
Comment (phracek) [inline rebasehelper/utils.py]: I guess, it should be logger.info. We should inform user, size of sources are different.

**Comment by phracek** [comment]
I guess, we used already urllib but for some reason, we remove it. PyCurl worked better then urllib.
Unfortunately, I don't understand why:(

**Comment by phracek** [comment]
You introduced DownloadHelperError, that's fine, but we should care about it here, right? https://github.com/phracek/rebase-helper/blob/master/rebasehelper/specfile.py#L174

File: rebasehelper/utils.py (hunk 1 above)
Comment (thozza) [inline rebasehelper/utils.py]: it works with ANY URL.

**Comment by thozza** [comment]
@phracek please check all commits.

File: rebasehelper/utils.py (hunk 2 above)
Comment (thozza) [inline rebasehelper/utils.py]: This is only to decide if the existing file is correct or not. e.g. when you interrupt rebase-helper while it is downloading a file, the file will stay there, but the sizes will be different. We could improve it to delete the file in case the download didn't finish and there is KeyboardInterrupt, but I would leave that for different patch set.

**Comment by phracek** [comment]
I see, I overlook RebaseHelperError.

**Comment by thozza** [comment]
@phracek BTW CoprHelper uses urllib anyway. If there are some real issues with urllib I'm all for resolving them, but if not I'm all for pycurl, as I don't think we actually used any of its features that are not already in urllib.

File: rebasehelper/utils.py (hunk 2 above)
Comment (thozza) [inline rebasehelper/utils.py]: @phracek but if you want to expose this message to the user I can change it. I don't have strong opinion about the log level of this message. :wink:

Comment (phracek) [inline rebasehelper/utils.py]: @thozza I would be glad if you can change it to info. THX

File: rebasehelper/utils.py [hunk 3]
```diff
@@ -202,22 +202,29 @@ def capture_output(func, capture_stdout=False, capture_stderr=False):
...
         :param url: URL from which to download the file
-        :param destination_name: path where to store downloaded file
+        :param destination_path: path where to store downloaded file
+        :param timeout: timeout in seconds for blocking actions like connecting, etc.
+        :param blocksize: size in Bytes of blocks used for downloading the file and reporting progress
         :return: None
         """
-        def progress(download_total, downloaded, upload_total, uploaded):
+        def progress(download_total, downloaded, start_time):
             r = downloaded / download_total if download_total else 0.0
```
Comment (nforro) [inline rebasehelper/utils.py]: `downloaded` or `download_total` (or both) have to be casted to *float*, otherwise the progress shows only 0% or 100% and no estimated time.

**Comment by nforro** [comment]
I'm fine with removing pycurl.
@thozza Could you remove it also from *docs/conf.py* and *doc-requirements.txt*?

File: rebasehelper/utils.py (hunk 3 above)
Comment (thozza) [inline rebasehelper/utils.py]: It works for me. @nforro did you try it?

Comment (nforro) [inline rebasehelper/utils.py]: Sorry, I forgot to mention the problem is with python2 only. With python3 it works as expected.

Comment (thozza) [inline rebasehelper/utils.py]: right, I just realized that it does not work with Python 2 and I was testing using Python 3. I will fix it...

**Comment by thozza** [comment]
@nforro I included your comments. As a bonus I added tests for the download progress function to make sure it we don't break it in the future.

**Comment by thozza** [comment]
@phracek I also changed the log message regarding the file size to INFO as you wanted.

**Comment by phracek** [comment]
Great thanks.
